Add mouse control style to match other, non-C&C, RTS games

dvr333 opened the pull request:
As title. Addresses #16669, but without breaking expectations from 3rd-gen C&C players.

dvr333 commented:
I'm not sure why the ftl files have every line changed, I'm guessing a line ending difference between how my editor is set up and what was there before, but I have no idea why they would be overridden

PunkPun commented:
> I'm not sure why the ftl files have every line changed, I'm guessing a line ending difference between how my editor is set up and what was there before, but I have no idea why they would be overridden

Your editor is not respecting newline format. You are doing CRLF, changing it from LF. Literally adding a CR character to every line break

dvr333 commented:
Fixed the line endings, wish I knew how to reliably squash commits without having to force push

PunkPun commented:
> Fixed the line endings, wish I knew how to reliably squash commits without having to force push

Force push is the thing you do with the rebase workflow

PunkPun commented:
This seems to just create more inconsistencies. All unit orders remain on RMB, but just attack move is special to be moved to LMB?

dvr333 commented:
> This seems to just create more inconsistencies. All unit orders remain on RMB, but just attack move is special to be moved to LMB?

A more general solution would be to have a TargetedUnitOrderGenerator that can be on left-click regardless of which button sends orders to units, and have the AttackMoveOrderGenerator inherit from that. Currently Attack Move is the only order that is targeted, as mods don't really have targeted unit abilities, but the convention in RTS games as a genre is to treat attack move as a targeted unit ability.

PunkPun commented:
What about Guard and other "abilities"?

dvr333 commented:
Good point, Guard is also a targeted order, and I rarely use it exactly because it being on right click is awkward for me. I'll set up the generalized TargetedUnitOrderGenerator when I get the chance, likely tomorrow (busy with work and Hallowe'en stuff today)

PunkPun commented:
Both left and right mouse buttons now activate attack move with no method of cancelling. Also please run out tests suite before pushing out a PR

https://github.com/OpenRA/OpenRA/wiki/Contributing

`make check` for code style, and `make test` for unit tests

dvr333 commented:
Right, sorry, I'm doing the tests now.

Can reproduce the cancel bug, fixed in next commit

PunkPun commented on OpenRA.Mods.Common/Orders/TargetedUnitOrderGenerator.cs:
```diff
@@ -0,0 +1,27 @@
+﻿using System.Collections.Generic;
```
lacks copyright header

dvr333 replied:
Fixed

PunkPun reviewed (commented):
Does it seem correct that basically all the commands in the command bar still get cancelled by left click orders?

dvr333 commented:
> Does it seem correct that basically all the commands in the command bar still get cancelled by left click orders?

Ah, yeah, fixed that, and the chronotank teleport. Did a sweep for everything that inherits from OrderGenerator and UnitOrderGenerator to make sure I didn't miss any others

Not sure if the setting should be a check box or part of the dropdown. I feel like the dropdown would make more sense, but I'm not sure if I'm the only one

dvr333 commented:
After some thought I've realized that the way I'm doing it, where order generators being "targeted orders" is opt-in rather than opt-out (or just no options besides the user-side ones), is over-engineered and already causing massive headaches, let alone when other mods with their own orders/traits with orders update and have to track down every order type that involves a 2-step order selection and targeting process. I'm re-designing it from bleed to go the other way, where there'll be a mode where Action is on left button, but the default UnitOrderGenerator.Order(function) and it's attendant support functions uses the right button, since that just works as expected without having to fiddle with every UnitOrderGenerator derived class

dvr333 committed:
Add input style to match other, non-C&C, RTS games

Uses left-click for targeted orders (attack move, guard, selecting movement types from the command bar, etc.) and right-click for contextual ones. Works like basically all non-C&C games

Removed GameSettings.UseClassicMouseStyle, as it is replaced by MouseControlStyle


Changed Minelayer and Chronotank teleport to UnitOrderGenerator

Fixes bug with Modern controls issuing a move order cancelling a mine field order

Better organizes all unit orders into a single order generator base class, rather than having their logic spread across multiple base classes, which required some hacks to get the control logic to match

PunkPun commented on OpenRA.Game/Settings.cs:
```diff
@@ -278,7 +280,7 @@ public class GameSettings
 		public int MouseScrollDeadzone = 8;
 
-		public bool UseClassicMouseStyle = false;
+		public bool UseClassicMouseStyle => MouseControl == MouseControlType.Classic;
```
We don't have settings migrations just yet. We should use the `UseClassicMouseStyle` setting if it's set to true

dvr333 replied:
You mean have setting MouseControl to MouseControlType.Classic also set UseClassicMouseStyle to true explicitly, rather than via the getter?

PunkPun replied:
Settings are loaded via magic. By turning this into a property you are making it so it doesn't get loaded anymore. In turn this means that users of "Modern" control style will now be dergraded to "Classic"

dvr333 replied:
but the relevant setting is now MouseControl, not UseClassicMouseStyle. I just didn't change that check everywhere because UseClassicMouseStyle is peppered all over the code since it was a boolean between Modern and Classic. With being an enum between 3 options UseClassicMouseStyle is now just there for the checks. I mean, I could rename it "IsUsingClassicMouseStyle" for clarity, but it's a convenience field at this point

dvr333 replied:
Fair enough, it was a convenience field at this point, but that's confusing, so I've gone and changed all the relevant UseClassicMouseStyle checks to gs.MouseControl == MouseControlType.Classic, with some convenience local functions or variables where the check gets repeated

PunkPun replied:
since we are not keeping the mouse preference, we should probably bump the introduction prompt `IntroductionVersion`

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/IntroductionPromptLogic.cs:
```diff
@@ -25,11 +26,13 @@ public class IntroductionPromptLogic : ChromeLogic
 
-		readonly string classic;
-		readonly string modern;
+		readonly Dictionary<MouseControlType, string> controlTypes;
```
this could just be an internal variable that's captured by the lambda

dvr333 replied:
Only for IntroductionPromptLogic? That's an easy enough change in IntroductionPromptLogic, but in InputSettingsLogic the controlTypes Dictionary is set in a different scope than the scope it's used in, so I'm not sure if you want that changed too.

PunkPun commented on OpenRA.Game/World.cs:
```diff
@@ -115,6 +115,7 @@ public Player RenderPlayer
 		public bool IsLoadingGameSave => OrderManager.NetFrameNumber <= OrderManager.GameSaveLastFrame;
 
 		public int GameSaveLoadingPercentage => OrderManager.NetFrameNumber * 100 / OrderManager.GameSaveLastFrame;
+		public bool UsingDefaultOrderGenerator { get; private set; }
```
this looks very wrong, what's happening here?

dvr333 replied:
For the "Generals" input mode, the key difference between orders on right click and orders on left click is that only the orders issued after selecting units (without specifying any other orders) or after World.CancelInputMode() use right-click. This way it just tracks explicitly whether the player has specified an order to be issued, or is just using the default, without having to assume that the type used for defaultUnitOrderType isn't ambiguous with any explicit orders' OrderGenerators

PunkPun reviewed (commented):
For guard both mous buttons activate it

Maybe minimap should also use left click?

dvr333 commented:
> For guard both mous buttons activate it
> 
> Maybe minimap should also use left click?

Fixed the Guard bug too,

I thought minimap was handling things properly (maybe it was before the redesign), will take a look at that tomorrow

dvr333 commented:
Minimap behaviour fixed, including some spurious selection cursors in Modern mode that was pre-existing (it was consistent with how RA3 works with right-click orders, but a bug in both cases)

PunkPun reviewed (commented):
I can cancel attack and guard by double clicking left mouse button. Hold down shift

Minimap still seems broken. You move on it with left mouse button, but left also activates support powers and orders units

dvr333 commented:
> I can cancel attack and guard by double clicking left mouse button. Hold down shift
> 
> Minimap still seems broken. You move on it with left mouse button, but left also activates support powers and orders units

I also notice a sound bug with shift+left-click, I'll look into that too.

The minimap behaviour is working as intended, or at least, working as intended given that I meant for the term "Generals" to be a shorthand for "not like C&C games" without actually saying it. Technically C&C Generals always has the left mouse button move the minimap, but you can only issue right-click orders through the minimap, and have to switch to right-click to target support powers. This is inconsistent between the map and minimap, and a downgrade for targeted orders, unlike the other two input modes. However, that's *just* Generals, every other RTS game I can find with a minimap will prioritize targeted orders over minimap movement if there is an order ready to be targeted, though some (like StarCraft 2) will switch right-click to moving the minimap in that case (which is easier to code than what is currently in the PR and an approach I took initially because of that)

PunkPun reviewed (commented):
<img width="440" height="452" alt="Screenshot 2025-12-03 at 13 12 30" src="https://github.com/user-attachments/assets/78340051-3e53-47b6-9b46-6fc0b14ddbfa" />

It's getting quite crowded 😅 

I'm wondering wether to consider ctrl / shift modifiers as entering order mode, and thus asking for making the left-click orders.

When holding shift, deploy cursor is no longer visible. Try MCV's or transport vehicles

dvr333 commented:
I personally find ctrl+RMB makes perfect sense as the way to have that work in this context, since it's still a one-click command, just with a key held (as opposed to using the command bar). If you're talking about having a separate key for the two-step approach (like how "Attack Ground" keys work in other RTS games) then that would be left click as that would be a key press (not hold) to change mode. Since Shift is meant to queue whatever orders are given it shouldn't change the click behaviour from an unqueued order at all.

I'll look into the cursor issue, getting that to work properly has been a bit tricky

PunkPun commented:
Maybe for description to take up less space, we could remove the light with support powers?

dvr333 commented:
Makes sense, it's always left click. I assume you're proposing that for every mouse control type, not just Generals?

What about merging "Select Units...", "Place Buildings..." and "Support Powers..." into a two-line thing, since they are always left click?

PunkPun commented:
Could merge "Place Buildings..." and "Support Powers...". However selection is very different from both

dvr333 commented:
Is it? Every mode uses "Select units using the <Left> mouse button" line as a separate line, even Classic. Merging just "Place Buildings..." and "Support Powers..." doesn't save any lines

PunkPun commented on OpenRA.Game/Settings.cs:
```diff
@@ -270,6 +271,7 @@ public class GameSettings
 		public int ViewportEdgeScrollMargin = 5;
 
 		public bool LockMouseWindow = false;
+		public MouseControlType MouseControl = MouseControlType.Classic;
```
By default `UseClassicMouseStyle` was false, so should we have Modern be the default?

dvr333 replied:
Makes sense

PunkPun commented on OpenRA.Game/Settings.cs:
```diff
@@ -278,7 +280,6 @@ public class GameSettings
 		public int SelectionDeadzone = 24;
 		public int MouseScrollDeadzone = 8;
 
-		public bool UseClassicMouseStyle = false;
```
Could we check for this value in the introduction prompt, and if it is set and `MouseControl` isn't, we'd use this value as the default in the prompt?

Note that there's no workflow implemented for this, the closest to such checking would be update rules, so it could be quite a bit of scope screep. Though we could just assume that `MouseControl` isn't set if it's the default, and only activate this behaviour when updating from introduction prompt version 0 or 1

dvr333 replied:
Can we check for this value in the introduction prompt? Without having it in GameSettings I'm only seeing getting into raw YAML parsing as the solution to this, and I'm not sure we have the settings.yaml handy in this scope

pchote replied:
A better approach is to add a custom `[LoadUsing()]` to the `MouseControl` field that parses `UseClassicMouseStyle` if it exists.

pchote replied:
The new settings system i'm working on will enable these kind of migrations without having to keep the deprecated settings implemented here - so lets go with the introduction prompt path and i'll clean up here afterwards.

PunkPun commented on OpenRA.Mods.Cnc/Traits/SupportPowers/AttackOrderPower.cs:
```diff
@@ -87,7 +87,7 @@ public class SelectAttackPowerTarget : OrderGenerator
 		{
 			// Clear selection if using Left-Click Orders
-			if (Game.Settings.Game.UseClassicMouseStyle)
+			if (Game.Settings.Game.MouseControl == MouseControlType.Classic)
```
do we even need such comments?

PunkPun commented on OpenRA.Mods.Cnc/Traits/SupportPowers/ChronoshiftPower.cs:
```diff
@@ -153,7 +153,7 @@ sealed class SelectChronoshiftTarget : OrderGenerator
 			{
 				// Clear selection if using Left-Click Orders
-				if (Game.Settings.Game.UseClassicMouseStyle)
+				if (Game.Settings.Game.MouseControl == MouseControlType.Classic)
```
here as well

PunkPun commented on OpenRA.Mods.Common/Orders/ForceModifiersOrderGenerator.cs:
```diff
@@ -29,8 +29,16 @@ public override IEnumerable<Order> Order(World world, CPos cell, int2 worldPixel
+					mi.Button = Game.Settings.Game.MouseButtonPreference.ContextualAction;
+				else if (mi.Button == Game.Settings.Game.MouseButtonPreference.Cancel && Game.Settings.Game.MouseControl != MouseControlType.Classic)
+					mi.Button = MouseButton.None;
+			}
```
This looks like quite the hack. Could it have "HACK:" added in front of the comment?

Tangent:
The command bar orders do not persist when holding down shift and I see `cancelOnFirstUse` is at fault. Do you think making it bypasable by shift-modifier would be fine?

dvr333 replied:
Added HACK

Re: Tangent: I added a check for the Shift modifier before canceling input mode

PunkPun commented on OpenRA.Mods.Common/Orders/GuardOrderGenerator.cs:
```diff
@@ -41,15 +41,18 @@ public override IEnumerable<Order> Order(World world, CPos cell, int2 worldPixel
-			var target = FriendlyGuardableUnits(world, mi).FirstOrDefault();
-			if (target == null)
-				yield break;
+			if (mi.Button == expectedButton)
```
maybe just inverse the if for better styling & less nesting

PunkPun commented on OpenRA.Mods.Common/Orders/UnitOrderGenerator.cs:
```diff
@@ -81,11 +81,13 @@ public virtual string GetCursor(World world, CPos cell, int2 worldPixel, MouseIn
 				useSelect = target.Type == TargetType.Actor && target.Actor.Info.HasTraitInfo<ISelectableInfo>() &&
-					(mi.Modifiers.HasModifier(Modifiers.Shift) || world.Selection.Actors.Count == 0);
+					((mi.Modifiers.HasModifier(Modifiers.Shift) && !InputOverridesSelection(world, worldPixel, mi)) ||
+					cursorOrder == null || world.Selection.Actors.Count == 0);
```
we generally try to put the cheapest checks (in terms of performance) first in the evaluation, I see `InputOverridesSelection` is quite heavy so it should be last

PunkPun commented on OpenRA.Mods.Common/Widgets/RadarWidget.cs:
```diff
@@ -294,13 +297,13 @@ public override string GetCursor(int2 pos)
 			var cursor = world.OrderGenerator.GetCursor(world, cell, worldPixel, mi);
-			if (cursor == null)
-				return worldDefaultCursor;
+			if (cursor == null || (Game.Settings.Game.MouseControl != MouseControlType.Classic && cursor == worldSelectCursor))
```
this seems like a really weird check, could it get an explanatory comment?

PunkPun commented on OpenRA.Mods.Common/Widgets/RadarWidget.cs:
```diff
@@ -313,14 +316,18 @@ public override bool HandleMouseInput(MouseInput mi)
 
+			var minimapMoveButton = world.UsingDefaultOrderGenerator || Game.Settings.Game.MouseControl != MouseControlType.OtherRTS ?
+				Game.Settings.Game.MouseButtonPreference.MinimapCameraMove :
+				Game.Settings.Game.MouseButtonPreference.ContextualAction;
```
This begs for refactoring. You added `MinimapCameraMove` for the sole purpose of being used here, and now you are adding checks in this place as well

dvr333 replied:
I think I'll ditch `MinimapCameraMove` and just have this check choose `MouseButton.Left` or `MouseButton.Right` when using the OtherRTS style here, since `MouseButtonPreference` doesn't know about `World`

PunkPun commented on OpenRA.Mods.Common/Widgets/RadarWidget.cs:
```diff
@@ -71,6 +73,7 @@ public RadarWidget(World world, WorldRenderer worldRenderer)
 			this.world = world;
 			this.worldRenderer = worldRenderer;
 
+			isMinimapMoving = false;
```
unnecesary asignment, false is already the default

PunkPun commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -101,15 +101,18 @@ public override bool HandleMouseInput(MouseInput mi)
 
-				dragStart = mousePos;
-				isDragging = true;
+				if (World.UsingDefaultOrderGenerator || useClassicMouseStyle)
```
this seems like a bug. You can no longer cancel by dragging your mouse in Modern control cheme. 

In OtherRTS not dragging seems more reasonable. But do we need to not keep dragging?

PunkPun reviewed (commented):
The UI now looks great.

I see there's no way to cancel waypoint mode, except for adding another order like enabling attack move. Also, the waypoint mode get's confirmed by right click?

dvr333 commented:
> The UI now looks great.
> 
> I see there's no way to cancel waypoint mode, except for adding another order like enabling attack move. Also, the waypoint mode get's confirmed by right click?

Waypoint mode is really weird, especially since everything it does is done better by just holding shift (that is, it doesn't produce waypoints for actions independent of units selected as it does in classic C&C games). Ideally it would be able to handle any order but that doesn't even work in the other modes, so right now it just handles contextual actions (matching the functionality of the other modes), cancelled by clicking on it again (which AFAICT is the old C&C UI flow), though also by choosing another order

PunkPun commented on OpenRA.Mods.Common/Widgets/Logic/Settings/InputSettingsLogic.cs:
```diff
@@ -64,7 +70,7 @@ Func<bool> InitPanel(Widget panel)
 			var mouseControlDropdown = panel.Get<DropDownButtonWidget>("MOUSE_CONTROL_DROPDOWN");
 			mouseControlDropdown.OnMouseDown = _ => ShowMouseControlDropdown(mouseControlDropdown, gs);
-			mouseControlDropdown.GetText = () => gs.UseClassicMouseStyle ? classic : modern;
+			mouseControlDropdown.GetText = () => controlTypes[gs.MouseControlStyle];
```
could we avoid crashing if control cheme is unset? I know that it shouldn't be common, but I reckon we should handle the case

PunkPun reviewed (commented):
On modern you can't activate support powers on minimap

> Waypoint mode is really weird, especially since everything it does is done better by just holding shift (that is, it doesn't produce waypoints for actions independent of units selected as it does in classic C&C games). Ideally it would be able to handle any order but that doesn't even work in the other modes, so right now it just handles contextual actions (matching the functionality of the other modes), cancelled by clicking on it again (which AFAICT is the old C&C UI flow), though also by choosing another order

it is weird. But as of this PR, it doesn't work like contextual and still uses right click orders. Cancelling should be done with right click no?

dvr333 commented:
> On modern you can't activate support powers on minimap

That's always been true. Did you want me to change that in this PR? If so, did you want it on right click (consistent with other minimap orders) or left click (consistent with support powers targeted on the main map, and AFAICT way easier to code, in a way that will remove the HACK in ForceModifiersOrderGenerator along with a bunch of other bodges and hacks, but does require either reflection shenanigans or making a property that should be get-only settable to otherwise work the way PR currently does. https://github.com/dvr333/OpenRA/tree/left-click-attack-move-expectedbutton if you want to check that out)?

> it is weird. But as of this PR, it doesn't work like contextual and still uses right click orders. Cancelling should be done with right click no?

Fair, it is a two-step order when you click on it, so I see the logic of it using left click

dvr333 commented:
Merged the experimental branch. It fixes all the problems listed, both the hacks and the support powers issue on Modern, and also reduces the amount of hardcoding, which seems to be a recent goal overall. I'll leave the commits unsquashed for now to allow review of the different approaches

pchote commented on OpenRA.Mods.Common/Orders/OrderGenerator.cs:
```diff
@@ -17,9 +17,10 @@ namespace OpenRA.Mods.Common.Orders
 {
 	public abstract class OrderGenerator : IOrderGenerator
 	{
+		public abstract MouseButton ExpectedButton { get; set; }
```
Can we please make this a readonly protected field that the implementations assign to from their constructor? This will save me from having to re-engineer this as part of removing the static `Game.Settings`.

dvr333 replied:
You mean as a backing field? This name and type has to be something the IOrderGenerator knows about in order to make sure everything can check World.OrderGenerator.ExpectedButton

pchote replied:
Implement `MouseButton IOrderGenerator.ExpectedButton => expectedButton;` as a getter for the backing field.

dvr333 replied:
so, two problems: 
1. Assignment to a protected readonly in implementation/derived class constructors for an abstract base class is not possible, and trying to solve that by hiding (and having every implementation have its own expectedButton field) leads to
2. Naming rules mean that OrderGenerator and its subclasses cannot have the same field, since the field name has to start with an uppercase letter in abstract classes but a lowercase letter in concrete classes.

pchote commented on OpenRA.Game/World.cs:
```diff
@@ -169,7 +168,14 @@ public IOrderGenerator OrderGenerator
+			OrderGenerator = (IOrderGenerator)modData.ObjectCreator.CreateBasic(defaultOrderGeneratorType);
+
+			// The default OrderGenerator's ExpectedButton property is set to ContextualAction here
+			// so that all mouse control styles work as expected in other mods by default
```
Breaking mods is preferrable to ugly code workarounds!

dvr333 replied:
ugh, fine. Having the default order generator work differently because it is the default (rather than because of its type) is the intended behaviour, but fine.

pchote commented:
Would it be possible to structure the PR with an initial commit that generalises the plumbing, and then a second commit to add the new mode and UI?

dvr333 commented:
> Would it be possible to structure the PR with an initial commit that generalises the plumbing, and then a second commit to add the new mode and UI?

Done

pchote commented on OpenRA.Game/Settings.cs:
```diff
@@ -19,6 +19,7 @@
 
 namespace OpenRA
 {
+	public enum MouseControlStyle { Unset, Classic, Modern, OtherRTS }
```
We should be able to drop `Unset` here?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/IntroductionPromptLogic.cs:
```diff
@@ -43,8 +44,20 @@ public IntroductionPromptLogic(Widget widget, ModData modData, WorldRenderer wor
+
+			if (gs.MouseControlStyle == MouseControlStyle.Unset)
+			{
+				// TODO: Have this check the now-removed UseClassicMouseStyle configuration option when IntroductionPromptVersion < 2
```
Once #22219 has been merged you can access this through something like `gs.Yaml.NodeWithKeyOrDefault("UseClassicMouseStyle");`

dvr333 replied:
Noted, I'll add a TODO

PunkPun commented:
Needs a rebase

dvr333 commented:
Rebased

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Settings/InputSettingsLogic.cs:
```diff
@@ -152,17 +164,18 @@ Action ResetPanel(Widget panel)
 		public static void ShowMouseControlDropdown(DropDownButtonWidget dropdown, GameSettings gameSettings)
 		{
-			var options = new Dictionary<string, bool>()
+			var options = new Dictionary<string, MouseControlStyle>()
```
This is duplicating `controlTypes`, and also keying on the button label is a bit odd. Not your fault, but could you please remove the `static` from this method and remove this in favour of `controlTypes`?

dvr333 replied:
I'm either keeping this or creating a reverse dictionary on the fly. If that's alright, cool, but that's what it needs

pchote replied:
Use this:
```csharp
public static void ShowMouseControlDropdown(DropDownButtonWidget dropdown, Dictionary<MouseControlStyle, string> controlTypes, GameSettings gameSettings)
{
	ScrollItemWidget SetupItem(MouseControlStyle o, ScrollItemWidget itemTemplate)
	{
		var item = ScrollItemWidget.Setup(itemTemplate,
			() => gameSettings.MouseControlStyle == o,
			() => gameSettings.MouseControlStyle = o);

		var label = controlTypes[o];
		item.Get<LabelWidget>("LABEL").GetText = () => label;
		return item;
	}

	dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 500, controlTypes.Keys, SetupItem);
}
```

It needs to stay static so it can be used by `IntroductionPromptLogic`.

dvr333 replied:
It either stays static or gets to use the controlTypes field, but not both

pchote replied:
Snippet above takes `controlTypes` as an argument, which is available from both places that call it.

dvr333 replied:
ah, sorry, missed that

pchote commented on OpenRA.Mods.Common/Widgets/Logic/IntroductionPromptLogic.cs:
```diff
@@ -43,8 +44,18 @@ public IntroductionPromptLogic(Widget widget, ModData modData, WorldRenderer wor
+			};
+
+			// TODO: Have this check gs.Yaml.NodeWithKeyOrDefault("UseClassicMouseStyle") when IntroductionPromptVersion < 2
+			// Requires PR #22219 to be merged
```
This has now been merged. If you change `IntroductionVersion` back to 1 I could implement and migrate the settings from #22253 in its own PR.

dvr333 replied:
This doesn't work at all if InputSettingsLogic.ShowMouseControlDropdown is non-static, since that's used here too

pchote replied:
Did you mean to reply to the other comment thread? Note that my suggestion there isn't static!

pchote commented on OpenRA.Mods.Common/Widgets/RadarWidget.cs:
```diff
@@ -303,8 +306,11 @@ public override string GetCursor(int2 pos)
-			if (cursor == null)
-				return worldDefaultCursor;
+
+			// We can't select through the minimap in Mouse Control Types other than Classic,
```
Can we please extract these other fixes to their own commits?

dvr333 replied:
Do you just want the cursor fixes extracted, or for all the changes to RadarWidget.cs to be in a different commit?

pchote replied:
One commit for each distinct feature/fix/change please, so I imagine these should each be their own?

> Fix erroneous minimap selection cursor when not in classic mode
>
> Fix adding 0 units to selection causing unit selected sounds to play
> 
> Fix selection cursor behaviour when not in classic mouse style
> 
> Fix cursor to always check if a given order would override selection before showing selection cursor, not just in Classic mode

dvr333 replied:
1, 3, and 4 would make sense as the same commit, but otherwise fair

dvr333 replied:
Done

pchote commented on OpenRA.Mods.Common/Widgets/WorldInteractionControllerWidget.cs:
```diff
@@ -86,7 +86,9 @@ public override bool HandleMouseInput(MouseInput mi)
 
-			var useClassicMouseStyle = gameSettings.UseClassicMouseStyle;
+			var useClassicMouseStyle = gameSettings.MouseControlStyle == MouseControlStyle.Classic;
+			var expectedButton = World.OrderGenerator.ActionButton;
```
`var actionButton`

dvr333 committed:
Fix playing sounds when not adding new units to selection

dvr333 committed:
Fix selection cursor behaviour

Fix erroneous minimap selection cursor when not in classic mode

Fix selection cursor not showing up with units selected when not in classic mouse style

dvr333 commented:
I'll squash the "Fix from review" commit to the first commit once it's been double-checked (Edit: is now squashed)

PunkPun reviewed (commented):
something has gone wrong with minelayer with modern control cheme. The order now persists when you hold down ctrl.

You also can't see the mine area after you've placed the order

Otherwise everything seems to work perfectly

dvr333 commented:
> something has gone wrong with minelayer with modern control cheme. The order now persists when you hold down ctrl.
> 
> You also can't see the mine area after you've placed the order
> 
> Otherwise everything seems to work perfectly

Bisecting a bit, the issue came up during the mouse input refactor, as the earlier version only worked by coincidence. Namely, OrderGenerator.Order() used to handle RMB on mouse up and LMB on mouse down, while now it (more correctly to the intent) handles CancelButton() on mouse up and ActionButton() on mouse down. The solutions I see are either doing the HasMouseFocus check on the right mouse button block in WorldInteractionControllerWidget.HandleMouseInput(), which is why this doesn't break with Classic controls as well; changing the way WorldInteractionControllerWidget.HandleMouseInput() works to RMB logic on mouse down instead of mouse up; or changing the Minefield order generator to be a UnitOrderGenerator instead of an OrderGenerator (along with PortableChrono, since that has the same bug). 

Changing to UnitOrderGenerator is my preferred solution, since I don't understand why these are inheriting from OrderGenerator in the first place

dvr333 commented:
I've fixed the issue by changing the base class for mineflelds and portablechrono order generators to UnitOrderGenerator, but in a separate commit in case that change is out of scope or not the desired solution

PunkPun commented:
Everything seems to be working correctly, just squash thge 1st and 4th commits

dvr333 commented:
Done

PunkPun reviewed (dismissed):
Alright

pchote commented on OpenRA.Mods.Common/Widgets/Logic/IntroductionPromptLogic.cs:
```diff
@@ -43,8 +44,18 @@ public IntroductionPromptLogic(Widget widget, ModData modData, WorldRenderer wor
+				{ MouseControlStyle.OtherRTS, FluentProvider.GetMessage(OtherRTS) },
+			};
+
+			if (gameSettings.Yaml.NodeWithKeyOrDefault("MouseControlStyle") == null &&
```
I suggested keeping `IntroductionVersion = 1` under the assumption that we'd do the setting migration in a followup PR - we do need to increment it if we're going to include this now!

IMO for future maintainability it would be better to explicitly gate this on the IntroductionPromptVersion, like this:
```csharp
if (gameSettings.IntroductionPromptVersion < 2)
{
	// UseClassicMouseStyle boolean was replaced by MouseControlStyle enum to support additional control styles
	var classicMouseStyleNode = gameSettings.Yaml.NodeWithKeyOrDefault("UseClassicMouseStyle");
	if (classicMouseStyleNode != null)
	{
		var useClassicMouseStyle = FieldLoader.GetValue<bool>("UseClassicMouseStyle", classicMouseStyleNode.Value.Value);
		gameSettings.MouseControlStyle = useClassicMouseStyle ? MouseControlStyle.Classic : MouseControlStyle.Modern;
	}
}
```

pchote commented on OpenRA.Mods.Common/Widgets/Logic/IntroductionPromptLogic.cs:
```diff
@@ -43,8 +44,18 @@ public IntroductionPromptLogic(Widget widget, ModData modData, WorldRenderer wor
+				bool.TryParse(gameSettings.Yaml.NodeWithKeyOrDefault("UseClassicMouseStyle")?.Value.Value ?? "False", out var useClassicMouseStyle))
+				gameSettings.MouseControlStyle = useClassicMouseStyle ? MouseControlStyle.Classic : MouseControlStyle.Modern;
+
+			bool UseClassicMouseStyle() => gameSettings.MouseControlStyle == MouseControlStyle.Classic;
```
This is only used twice, so isn't worth being its own function. Can we please inline this below?

dvr333 commented:
pchote's requested fixes squashed into the first commit

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ce3418fe9b6cdcd081436ea1fb9ddb3c27e61bb3
